### PR TITLE
[SYCL][HIP] Support of AMD matrix core instructions

### PR DESCRIPTION
**Pull request by mmoadeli**
* Support one block AMD matrix core instructions for `__gfx90a__ architecture.
* Supports `__builtin_amdgcn_mfma_i32_32x32x8i8`, `__builtin_amdgcn_mfma_i32_16x16x16i8`, `__builtin_amdgcn_mfma_f64_16x16x4f64`, `__builtin_amdgcn_mfma_f32_32x32x8bf16_1k`, `__builtin_amdgcn_mfma_f32_16x16x16bf16_1k`, `__builtin_amdgcn_mfma_f32_32x32x8f16` and `__builtin_amdgcn_mfma_f32_16x16x16f16` instructions.

### CODE REVIEW

File: sycl/include/sycl/ext/oneapi/matrix/matrix-hip.hpp [hunk 1]
```diff
@@ -0,0 +1,317 @@
...
+    joint_matrix_hip<S, matrix_use::accumulator, NumRows, NumCols,
+                     matrix_layout::dynamic> &res,
+    multi_ptr<T, Space, IsDecorated> src, size_t stride, Group &sg) {
+  std::ignore = stride;
+  const auto idx = sg.get_group_linear_id() * sg.get_local_range()[0] +
+                   sg.get_local_linear_id();
+  if constexpr (Layout == matrix_layout::row_major) {
+    res.data = {0};
+  } else if constexpr (Layout == matrix_layout::col_major) {
+    res.data = {0};
```
Comment (JackAKirk) [inline sycl/include/sycl/ext/oneapi/matrix/matrix-hip.hpp]: So it seems this doesn't load any data to the accumulator? I guess you'll add this before opening a PR?

File: sycl/include/sycl/ext/oneapi/matrix/matrix-hip.hpp [hunk 2]
```diff
@@ -0,0 +1,317 @@
...
+  }
+}
+
+template <
+    typename Group, typename S, typename T, size_t NumRows, size_t NumCols,
+    matrix_use Use, matrix_layout Layout, access::address_space Space,
+    access::decorated IsDecorated,
+    typename = typename std::enable_if_t<(Layout == matrix_layout::row_major ||
+                                          Layout == matrix_layout::col_major)>>
+void load_multiplicand_hip(
```
Comment (JackAKirk) [inline sycl/include/sycl/ext/oneapi/matrix/matrix-hip.hpp]: Have you tried loading from const matrix? I'm guessing the type checks will break it below?

Might have to do something similar to https://github.com/intel/llvm/pull/6532/files

File: sycl/include/sycl/ext/oneapi/matrix/matrix-hip.hpp [hunk 3]
```diff
@@ -0,0 +1,317 @@
...
+  std::ignore = stride;
+  const auto idx = sg.get_group_linear_id() * sg.get_local_range()[0] +
+                   sg.get_local_linear_id();
+
+  if constexpr (std::is_same_v<T, double>) {
+    if constexpr (NumRows == 16 && NumCols == 16) {
+      const auto thread_x = idx % NumCols;
+      const auto thread_y = idx / NumCols;
+      for (int i = 0; i < 4; ++i) {
+        dst[thread_x + i * 4 * NumCols + thread_y * NumCols] = src.data[i];
```
Comment (JackAKirk) [inline sycl/include/sycl/ext/oneapi/matrix/matrix-hip.hpp]: I wonder if this loop gets unrolled automatically or if not whether it affects perf?

File: sycl/include/sycl/ext/oneapi/matrix/matrix-hip.hpp [hunk 4]
```diff
@@ -0,0 +1,317 @@
...
+      for (int j = 0; j < 4; ++j) {
+        for (int i = 0; i < 4; ++i) {
+          const int d_idx =
+              thread_x + i * NumCols + thread_y * 4 * NumCols + j * 8 * NumCols;
+          dst[d_idx] = src.data[i + 4 * j];
+        }
+      }
+    }
+  } else {
+    static_assert(false && "Invalid dimenstions!");
```
Comment (JackAKirk) [inline sycl/include/sycl/ext/oneapi/matrix/matrix-hip.hpp]: Is it possible to define a joint_matrix that can lead to this assert? Ideally it invalid types/sizes can't be instantiated, such that asserts like these aren't required.

File: sycl/include/sycl/ext/oneapi/matrix/matrix-hip.hpp [hunk 5]
```diff
@@ -0,0 +1,317 @@
...
+    Group &sg) {
+  switch (Layout) {
+  case matrix_layout::row_major:
+    store_layoutT<Group, matrix_layout::row_major>(src, dst, stride, sg);
+    break;
+  case matrix_layout::col_major:
+    store_layoutT<Group, matrix_layout::col_major>(src, dst, stride, sg);
+    break;
+  default:
+    assert(false && "Invalid layout specified!");
```
Comment (JackAKirk) [inline sycl/include/sycl/ext/oneapi/matrix/matrix-hip.hpp]: I think only row_major or col_major layouts can be specified? Maybe I am wrong, but if this is the case such asserts aren't required.

File: sycl/include/sycl/ext/oneapi/matrix/matrix-hip.hpp (hunk 2 above)
Comment (mmoadeli) [reply sycl/include/sycl/ext/oneapi/matrix/matrix-hip.hpp]: The type checks have been on `S` which is the type joint_matrix created from. 
Even if `T`(multi_ptr type) is const, it should be fine, as the `const` value is assigned to non-const value of `joint_matrix` data.

File: sycl/include/sycl/ext/oneapi/matrix/matrix-hip.hpp (hunk 3 above)
Comment (mmoadeli) [reply sycl/include/sycl/ext/oneapi/matrix/matrix-hip.hpp]: That's fair point. I'd suggest this to investigated it later.

File: sycl/doc/extensions/experimental/sycl_ext_matrix/sycl_ext_oneapi_matrix.asciidoc [hunk 6]
```diff
@@ -964,6 +964,35 @@ multiple of 4 when `T` is `float`; where `T` is the type of the
...
 no restrictions to `stride`.
 
+==== AMD Matrix Cores Supported Combinations
+The complete set of matrix data types and dimenstions that are supported by
+the `ext_oneapi_hip` backend are represented in the following
+table. In this architecture's implementation, A and B matrices must have the same type. 
+Similarly, C and D matrices must share the same type.
+
+IMPORTANT: Currently, only one block AMD Matrix Core instructions in GFX90A and 
+GFX908 (MI200, MI210, MI250 and MI250X GPUs) architecture are supported.
```
Comment (JackAKirk) [inline sycl/doc/extensions/experimental/sycl_ext_matrix/sycl_ext_oneapi_matrix.asciidoc]: can you specify which are GFX908 and which are GFX90a?

**Pull request closed without merging**